File: app/src/main/java/app/simple/buyer/util/DBMigration.java
package app.simple.buyer.util;

import io.realm.DynamicRealm;
import io.realm.RealmMigration;
import io.realm.RealmSchema;

/**
 * Created by Zakharovi on 23.01.2018.
 */

public class DBMigration implements RealmMigration {
    @Override
    public void migrate(DynamicRealm realm, long oldVersion, long newVersion) {

        RealmSchema schema = realm.getSchema();
        if (oldVersion == 0) {
            schema.get("BuyItem")
                    .removeField("count");
            schema.get("BuyListItem")
                    .removeField("price");
            oldVersion++;
        }
//
//        // DynamicRealm exposes an editable schema
//
//
//        // Migrate to version 1: Add a new class.
//        // Example:
//        // public Person extends RealmObject {
//        //     private String name;
//        //     private int age;
//        //     // getters and setters left out for brevity
//        // }
//        if (oldVersion == 0) {
//            schema.create("Person")
//                    .addField("name", String.class)
//                    .addField("age", int.class);
//            oldVersion++;
//        }
//
//        // Migrate to version 2: Add a primary key + object references
//        // Example:
//        // public Person extends RealmObject {
//        //     private String name;
//        //     @PrimaryKey
//        //     private int age;
//        //     private Dog favoriteDog;
//        //     private RealmList<Dog> dogs;
//        //     // getters and setters left out for brevity
//        // }
//        if (oldVersion == 1) {
//            schema.get("Person")
//                    .addField("id", long.class, FieldAttribute.PRIMARY_KEY)
//                    .addRealmObjectField("favoriteDog", schema.get("Dog"))
//                    .addRealmListField("dogs", schema.get("Dog"));
//            oldVersion++;
//        }
    }
}

File: app/src/main/java/app/simple/buyer/util/database/DBMigration.kt
package app.simple.buyer.util.database

import app.simple.buyer.entities.BuyItem
import app.simple.buyer.entities.BuyList
import app.simple.buyer.entities.BuyListItem
import app.simple.buyer.entities.User
import io.realm.DynamicRealm
import io.realm.FieldAttribute
import io.realm.RealmMigration
import java.util.Date


/**
 * Created by Zakharovi on 23.01.2018.
 */

class DBMigration : RealmMigration {
    companion object {
        const val schemaVersion: Long = 11
    }

    override fun migrate(realm: DynamicRealm, oldVersion: Long, newVersion: Long) {
        val schema = realm.schema
        for (version in oldVersion..newVersion) when (version) {
            0L -> {
//                schema.get("BuyItem")?.removeField("count")
//                schema.get("BuyList")?.addField("isHidden", Boolean::class.java)
            }
            4L -> {
                schema.get(BuyItem.KEY_TABLE_NAME)?.renameField("populatity", BuyItem.KEY_POPULARITY)
                schema.get(BuyList.KEY_TABLE_NAME)?.renameField("populatity", BuyList.KEY_POPULARITY)
            }
            5L -> {
                if (schema.get(User.KEY_TABLE_NAME)?.hasField(User.KEY_SHOW_CHECKED_ITEMS) == false) {
                    schema.get(User.KEY_TABLE_NAME)?.addField(User.KEY_SHOW_CHECKED_ITEMS, Boolean::class.java)
                }
            }
            6L -> {
                schema.get(BuyListItem.KEY_TABLE_NAME)?.addField(BuyListItem.KEY_IS_SELECTED, Boolean::class.java)
            }
            7L -> {
                schema.get(BuyListItem.KEY_TABLE_NAME)?.addField(BuyListItem.KEY_IS_ARCHIVED, Boolean::class.java)
            }
            8L -> {
                schema.get(BuyItem.KEY_TABLE_NAME)?.addField(BuyItem.KEY_UNIT, Int::class.java)
                schema.get(BuyListItem.KEY_TABLE_NAME)?.addField(BuyListItem.KEY_SCROLL_STATE, ByteArray::class.java, FieldAttribute.REQUIRED)
                schema.get(User.KEY_TABLE_NAME)?.addField(User.KEY_CURRENT_ITEM_ID, Long::class.java)
            }
            10L -> {
                schema.get(BuyListItem.KEY_TABLE_NAME)?.addField(BuyListItem.KEY_BUYED, Date::class.java, FieldAttribute.REQUIRED)
            }
            11L -> {
                realm.executeTransactionAsync {
                    val tt = it.where(BuyListItem::class.java.name).findAll()
                    tt.forEach { buyListItem ->
                        buyListItem.set(BuyListItem.KEY_BUYED, buyListItem.get(BuyListItem.KEY_MODIFIED))
                    }
                }
            }
        }
    }
}